src/responder/kcm/kcmsrv_ccache_secdb.c

```diff
@@ -0,0 +1,1426 @@
+    return ret;
+}
+
+static errno_t key_by_name(TALLOC_CTX *mem_ctx,
+                           struct sss_sec_ctx *sctx,
+                           struct cli_creds *client,
+                           const char *name,
+                           char **_key)
+{
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+    char *key_match = NULL;
+    char **keys = NULL;
+    size_t nkeys;
+    struct sss_sec_req *sreq = NULL;
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    ret = secdb_container_url_req(tmp_ctx, sctx, client, &sreq);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = sss_sec_list(tmp_ctx, sreq, &keys, &nkeys);
+    if (ret == ENOENT) {
+        DEBUG(SSSDBG_MINOR_FAILURE, "The container was not found\n");
+        goto done;
+    } else if (ret != EOK) {
+        goto done;
+    }
+
+    for (size_t i = 0; i < nkeys; i++) {
+        if (sec_key_match_name(keys[i], name)) {
+            key_match = keys[i];
+            break;
+        }
+    }
+
+    if (key_match == NULL) {
+        ret = ENOENT;
+        goto done;
+    }
+
+    ret = EOK;
+    *_key = talloc_steal(mem_ctx, key_match);
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t secdb_get_cc(TALLOC_CTX *mem_ctx,
+                            struct sss_sec_ctx *sctx,
+                            const char *secdb_key,
+                            struct cli_creds *client,
+                            struct kcm_ccache **_cc)
+{
+    errno_t ret;
+    TALLOC_CTX *tmp_ctx = NULL;
+    struct kcm_ccache *cc = NULL;
+    struct sss_sec_req *sreq = NULL;
+    struct sss_iobuf *ccbuf;
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    ret = secdb_cc_key_req(tmp_ctx, sctx, client, secdb_key, &sreq);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = sec_get_b64(tmp_ctx, sreq, &ccbuf);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = sec_kv_to_ccache(tmp_ctx,
+                           secdb_key,
+                           (const char *) sss_iobuf_get_data(ccbuf),
+                           client,
+                           &cc);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Cannot convert JSON keyval to ccache blob [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    ret = EOK;
+    *_cc = talloc_steal(mem_ctx, cc);
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t ccdb_secdb_init(struct kcm_ccdb *db)
+{
+    struct ccdb_secdb *secdb = NULL;
+    errno_t ret;
+
+    secdb = talloc_zero(db, struct ccdb_secdb);
+    if (secdb == NULL) {
+        return ENOMEM;
+    }
+
+    /* TODO: read configuration from the config file, adjust quotas */
+
+    ret = sss_sec_init(db, NULL, &secdb->sctx);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Cannot initialize the security database\n");
+        return ret;
```

Comment:
Aren't you leaking secdb here?
I mean, it's under db and will be freed when db is freed, but wouldn't be better/cleaner to explicitly free it here?

Reply:
You're right, it's cleaner